tests/app/one_click_unsubscribe/test_one_click_unsubscribe.py

```diff
@@ -0,0 +1,116 @@
+import uuid
+
+from flask import current_app
+from notifications_utils.url_safe_token import generate_token
+
+from app.constants import EMAIL_TYPE
+from app.dao.templates_dao import dao_update_template
+from app.dao.unsubscribe_request_dao import get_unsubscribe_request_by_notification_id_dao
+from tests.app.db import create_notification, create_template
+
+
+def unsubscribe_url_post(client, notification_id, token):
+    return client.post(path=f"/unsubscribe/{notification_id}/{token}")
+
+
+def test_valid_one_click_unsubscribe_url(client, sample_email_notification):
+    token = generate_token(
+        sample_email_notification.to, current_app.config["SECRET_KEY"], current_app.config["DANGEROUS_SALT"]
+    )
+    response = unsubscribe_url_post(client, sample_email_notification.id, token)
+    response_json_data = response.get_json()
+    created_unsubscribe_request = get_unsubscribe_request_by_notification_id_dao(sample_email_notification.id)
+    assert response.status_code == 200
+    assert response_json_data["message"] == "Unsubscribe successful"
+    assert response_json_data["result"] == "success"
+    assert created_unsubscribe_request.notification_id == sample_email_notification.id
+    assert created_unsubscribe_request.template_id == sample_email_notification.template_id
+    assert created_unsubscribe_request.template_version == sample_email_notification.template_version
+    assert created_unsubscribe_request.service_id == sample_email_notification.service_id
+    assert created_unsubscribe_request.email_address == sample_email_notification.to
+
+
+def test_unsubscribe_request_object_refers_to_correct_template_version_after_template_updated(client, sample_service):
+    test_template = create_template(
+        sample_service,
+        template_type=EMAIL_TYPE,
+    )
+    notification = create_notification(template=test_template, to_field="foo@bar.com")
+    initial_template_version = test_template.version
+    token = generate_token(notification.to, current_app.config["SECRET_KEY"], current_app.config["DANGEROUS_SALT"])
+
+    response = unsubscribe_url_post(client, notification.id, token)
+
+    # update template content to generate new template version
+    test_template.content = "New content"
+    test_template.process_type = "priority"
+    dao_update_template(test_template)
```

Comment:
should this step should happen before we call `unsubscribe_url_post`?

so in order:

* create notification
* update template
* unsubscribe endpoint is called

Reply:
Thanks for catching that. It was a mistake and I have corrected the order.